pkg/config/config.go

```diff
@@ -37,38 +37,81 @@ var (
 	validRoles = []string{"controlplane", "node"}
 )
 
-type ClusterConfig struct {
-	URL string `json:"url"`
-
-	ClusterCIDR          string `json:"clusterCIDR"`
-	ServiceCIDR          string `json:"serviceCIDR"`
-	ServiceNodePortRange string `json:"serviceNodePortRange"`
-	DNS                  string `json:"dns"`
-	Domain               string `json:"domain"`
-	MTU                  string `json:"mtu"`
+// Top level config
+type Config struct {
+	Components Components       `json:components`
+	Microshift MicroshiftConfig `json:microshift`
 }
 
-type DebugConfig struct {
-	Pprof bool `json:"pprof"`
+// Components config
+type Components struct {
+	Network   Network   `json:network`
+	DNS       DNS       `json:dns`
+	APIServer APIServer `json:apiServer`
 }
 
-type MicroshiftConfig struct {
-	ConfigFile string `json:"configFile"`
-	DataDir    string `json:"dataDir"`
+type Network struct {
+	// IP address pool to use for pod IPs.
+	// This field is immutable after installation.
+	ClusterNetwork []ClusterNetworkEntry `json:"clusterNetwork"`
+
+	// IP address pool for services.
+	// Currently, we only support a single entry here.
+	// This field is immutable after installation.
+	ServiceNetwork []string `json:"serviceNetwork"`
+
+	// The port range allowed for Services of type NodePort.
+	// If not specified, the default of 30000-32767 will be used.
+	// Such Services without a NodePort specified will have one
+	// automatically allocated from this range.
+	// This parameter can be updated after the cluster is
+	// installed.
+	// +kubebuilder:validation:Pattern=`^([0-9]{1,4}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])-([0-9]{1,4}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])$`
+	ServiceNodePortRange string `json:"serviceNodePortRange,omitempty"`
+
+	// Note: additional field to config v1 Network type
+	MTU string `json:"mtu"`
```

Comment:
let's model it like

```go
type Network struct{
  ClusterNetwork []ClusterNetworkEntry`json:"clusterNetwork"`
  ServiceNetwork []string`json:"serviceNetwork"`

  MicroshiftNetwork `json:",inline"`
}

type MicroshiftNetwork struct{
  MTU string `json:"mtu"`
}
```

doing this will keep the region segregated and make the additional fields clearly opt-in at the code level while keeping the serialization looking like

```yaml
network:
  serviceNetwork: valueOne
  mtu: valueHere
```

Reply:
Agree that the `mtu` field shall be moved out from the network main spec to align with openshift [network.config](https://github.com/openshift/api/blob/master/config/v1/types_network.go#L34) spec.
This `mtu` value is correlated to CNI implementation. For ovn-kubernetes, it is the tunnel interface mtu, instead of the uplink mtu. It would be good to move it to the CNI specific struct, such as OVNKubernetesConfig, just like how it is defined in the [network operator API](https://github.com/openshift/api/blob/master/operator/v1/types_network.go#L421). This puts mtu in the context of ovn-kubernetes CNI and should help to reduce the confusion.

Perhaps the `MicroshiftNetwork` can be renamed to `OVNKubernetesNetwork` for our current CNI choice. Other CNI network struct (if there will be) can be simply appended to the `Network` struct.

Reply:
@zshi-redhat makes a good point. If the MTU value is only used by CNI, we may want that in a different config file entirely. We have all of the TopoLVM config in a separate file, for example, and that's going to make it much easier to decouple that work so the community can swap in different CSI implementations. We will want to be able to do the same for CNI.

Reply:
Thank you all for stepping up to discuss, it took a direction we didn't expect, but it sounds better than the original proposal.

My question is should we really decouple that from the microshift configuration? Having everything in one place would make it easier for the users IMO. We could have dedicated entries in the config for `CNIPlugins` and `CSIPlugins`, with a similar layout as what we have today in the operator configs: https://github.com/openshift/api/blob/4226c2167e40c62b09e2db083d1c0051fafba7da/operator/v1/types_network.go#L196-L212

Reply:
David's suggestion sounds good to me. Will make the changes for the struct changes.